python/taichi/lang/snode.py

```diff
@@ -48,16 +51,25 @@ def place(self, *args, offset=None):
     def lazy_grad(self):
         self.ptr.lazy_grad()
 
-    def parent(self):
-        return SNode(self.ptr.snode().parent)
+    def parent(self, n=1):
+        impl.get_runtime().try_materialize()
+        p = self.ptr
+        for i in range(n):
+            p = p.parent
+        if p.type == impl.taichi_lang_core.SNodeType.root:
```

Comment:
IIRC, the parent of `root` is `nullptr`, so I think we should return `None` instead of making a cycle?